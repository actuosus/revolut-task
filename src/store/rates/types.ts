import { Currency } from "../wallets/types";

export const GET_RATES = "GET_RATE";
export const GET_RATES_REQUEST = "GET_RATE_REQUEST";
export const GET_RATES_SUCCESS = "GET_RATE_SUCCESS";
export const GET_RATE_FAILURE = "GET_RATE_FAILURE";
export const SUBSCRIBE_TO_RATES = "SUBSCRIBE_TO_RATES";
export const UNSUBSCRIBE_FROM_RATES = "UNSUBSCRIBE_FROM_RATES";

export type Rate = Partial<Record<Currency, number>>;

export type RatesState = Partial<Record<Currency, Rate>> & {
  error?: string | null;
};

export interface RatesData {
  currency: Currency;
  rates: Partial<Record<Currency, number>>;
}

export interface GetRatesAction {
  type: typeof GET_RATES;
  payload: {
    currency: Currency;
    interval?: number;
  };
}

export interface GetRatesRequestAction {
  type: typeof GET_RATES_REQUEST;
}

export interface GetRatesSuccessAction {
  type: typeof GET_RATES_SUCCESS;
  payload: {
    currency: Currency;
    rates: Partial<Record<Currency, number>>;
  };
}

export interface GetRatesFailureAction {
  type: typeof GET_RATE_FAILURE;
  payload: Error;
}

export interface SubscribeToRatesAction {
  type: typeof SUBSCRIBE_TO_RATES;
  payload: {
    currency: Currency;
    interval?: number;
  };
}

export interface UnsubscribeFromRatesAction {
  type: typeof UNSUBSCRIBE_FROM_RATES;
  payload: {
    currency: Currency;
  };
}

export type RatesActions =
  | GetRatesAction
  | GetRatesRequestAction
  | GetRatesSuccessAction
  | GetRatesFailureAction
  | SubscribeToRatesAction
  | UnsubscribeFromRatesAction;
