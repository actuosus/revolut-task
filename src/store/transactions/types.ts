import { WalletId } from "../wallets/types";

export type TransactionId = string;
export enum TransactionType {
  TRANSFER = 10
}
export interface BaseTransaction {
  id: TransactionId;
  type: TransactionType;
  timestamp: number;
}

export interface TransferTransaction extends BaseTransaction {
  fromWalletId: WalletId;
  toWalletId: WalletId;
  amount: number;
  fromAmount: number;
}

export type Transaction = TransferTransaction;
export type TransactionsState = Record<TransactionId, Transaction>;

export type TransferAction = {
  type: "TRANSFER";
  payload: {
    fromWalletId: WalletId;
    toWalletId: WalletId;
    amount: number;
    rate: number;
    timestamp: number;
  };
};

export type TransactionActions = TransferAction;
