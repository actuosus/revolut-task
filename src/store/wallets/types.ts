export type Currency =
  | "EUR"
  | "USD"
  | "JPY"
  | "BGN"
  | "CZK"
  | "DKK"
  | "GBP"
  | "HUF"
  | "PLN"
  | "RON"
  | "SEK"
  | "CHF"
  | "ISK"
  | "NOK"
  | "HRK"
  | "RUB"
  | "TRY"
  | "AUD"
  | "BRL"
  | "CAD"
  | "CNY"
  | "HKD"
  | "IDR"
  | "ILS"
  | "INR"
  | "KRW"
  | "MXN"
  | "MYR"
  | "NZD"
  | "PHP"
  | "SGD"
  | "THB"
  | "ZAR";
export type WalletId = string;
export interface Wallet {
  id: WalletId;
  currency: Currency;
  balance: number;
}

export type WalletsState = Record<WalletId, Wallet>;

export const EXCHANGE = "EXCHANGE";
export const EXCHANGE_REQUEST = "EXCHANGE_REQUEST";
export const EXCHANGE_SUCCESS = "EXCHANGE_SUCCESS";
export const EXCHANGE_FAILURE = "EXCHANGE_FAILURE";

export interface ExchangeAction {
  type: typeof EXCHANGE;
  payload: {
    fromWalletId: WalletId;
    toWalletId: WalletId;
    amount: number;
    rate: number;
    timestamp: number;
  };
}

export type WalletActions = ExchangeAction;
