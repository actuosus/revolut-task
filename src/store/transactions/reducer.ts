import {
  TransactionActions,
  TransactionsState,
  TransactionType
} from "./types";

const initialState: TransactionsState = {
  "1": {
    id: "1",
    type: TransactionType.TRANSFER,
    fromWalletId: "2",
    toWalletId: "1",
    amount: 50,
    fromAmount: 37.18,
    timestamp: 1452261660000 // 8 Jan 2016
  },
  "2": {
    id: "2",
    type: TransactionType.TRANSFER,
    fromWalletId: "1",
    toWalletId: "2",
    amount: 1.34,
    fromAmount: 1,
    timestamp: 1452261000000 // 8 Jan 2016
  },
  "3": {
    id: "3",
    type: TransactionType.TRANSFER,
    fromWalletId: "2",
    toWalletId: "1",
    amount: 7.45,
    fromAmount: 5,
    timestamp: 1452081600000 // 6 Jan 2016
  },
  "4": {
    id: "4",
    type: TransactionType.TRANSFER,
    fromWalletId: "2",
    toWalletId: "1",
    amount: 7.45,
    fromAmount: 5,
    timestamp: 1450459620000 // 18 Dec 2015
  },
  "5": {
    id: "5",
    type: TransactionType.TRANSFER,
    fromWalletId: "1",
    toWalletId: "3",
    amount: 7.45,
    fromAmount: 5,
    timestamp: 1450353600000 // 17 Dec 2015
  }
};

export const transactionReducer = (
  state = initialState,
  action: TransactionActions
) => {
  switch (action.type) {
    case "TRANSFER": {
      const lastTransactionId = Object.keys(state)
        .sort((a, b) => {
          if (parseInt(a) > parseInt(b)) {
            return 1;
          } else {
            return -1;
          }
        })
        .pop();
      const lastTransactionIdNumber = parseInt(lastTransactionId || "1");
      const newId = (lastTransactionIdNumber + 1).toString();

      return {
        ...state,
        [newId]: {
          id: newId,
          type: TransactionType.TRANSFER,
          fromWalletId: action.payload.fromWalletId,
          toWalletId: action.payload.toWalletId,
          amount: action.payload.amount * action.payload.rate,
          fromAmount: action.payload.amount,
          timestamp: action.payload.timestamp
        }
      };
    }

    default:
      return state;
  }
};

export default transactionReducer;
