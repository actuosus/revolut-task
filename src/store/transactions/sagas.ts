import { delay, put, takeEvery } from "redux-saga/effects";
import { EXCHANGE, ExchangeAction } from "../wallets/types";

function* recordTransferTransaction(action: ExchangeAction) {
  yield delay(1000);
  yield put({ type: "TRANSFER", payload: action.payload });
}

export function* watchExchange() {
  yield takeEvery(EXCHANGE, recordTransferTransaction);
}
