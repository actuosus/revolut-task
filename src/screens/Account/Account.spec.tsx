import { shallow, ShallowWrapper } from "enzyme";
import React from "react";
import configureStore from "redux-mock-store";
import { TransactionType } from "../../store/transactions/types";
import ConnectedAccount, { Account } from "./Account";

jest.mock("navigation");

describe("<Account />", () => {
  const initialState = {
    wallets: {
      1: { id: 1, currency: "USD", balance: 10 },
      2: { id: 2, currency: "GBP", balance: 12 }
    },
    transactions: {
      1: {
        type: TransactionType.TRANSFER,
        id: 1,
        fromWalletId: 1,
        toWalletId: 2,
        amount: 10,
        fromAmount: 2
      }
    }
  };
  const mockStore = configureStore();
  let store, wrapper: ShallowWrapper, component: ShallowWrapper, navigation;

  beforeEach(() => {
    store = mockStore(initialState);
    navigation = require("navigation");
    wrapper = shallow(
      // @ts-ignore shallow wrapper with store
      <ConnectedAccount store={store} navigation={navigation} />
    );
    component = wrapper.find(Account);
  });

  it("matches its snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("has wallets", () => {
    expect(component.prop("wallets")).toEqual(initialState.wallets);
  });

  it("has transactions", () => {
    expect(component.prop("transactions")).toEqual(initialState.transactions);
  });
});
