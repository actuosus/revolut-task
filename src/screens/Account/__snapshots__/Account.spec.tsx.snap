// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Account /> matches its snapshot 1`] = `
<ContextProvider
  value={null}
>
  <Account
    dispatch={[Function]}
    navigation={
      Object {
        "default": Object {
          "addListener": [Function],
          "closeDrawer": [Function],
          "dangerouslyGetParent": [Function],
          "dismiss": [Function],
          "dispatch": [Function],
          "getParam": [Function],
          "goBack": [Function],
          "isFocused": [Function],
          "navigate": [Function],
          "openDrawer": [Function],
          "pop": [Function],
          "popToTop": [Function],
          "push": [Function],
          "replace": [Function],
          "setParams": [Function],
          "state": Object {
            "index": 0,
            "isTransitioning": false,
            "key": "",
            "params": Object {
              "fromWalletId": "1",
            },
            "routes": Array [],
          },
          "toggleDrawer": [Function],
        },
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    transactions={
      Object {
        "1": Object {
          "amount": 10,
          "fromAmount": 2,
          "fromWalletId": 1,
          "id": 1,
          "toWalletId": 2,
          "type": 10,
        },
      }
    }
    wallets={
      Object {
        "1": Object {
          "balance": 10,
          "currency": "USD",
          "id": 1,
        },
        "2": Object {
          "balance": 12,
          "currency": "GBP",
          "id": 2,
        },
      }
    }
  />
</ContextProvider>
`;
